src/core/dvr.c

```diff
@@ -0,0 +1,142 @@
+#include "dvr.h"
+
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include <log/log.h>
+#include <minIni.h>
+
+#include "core/msp_displayport.h"
+#include "driver/hardware.h"
+#include "ui/page_common.h"
+
+bool dvr_is_recording = false;
+
+static pthread_mutex_t dvr_mutex;
+
+///////////////////////////////////////////////////////////////////
+//-1=error;
+// 0=idle,1=recording,2=stopped,3=No SD card,4=recorf file path error,
+// 5=SD card Full,6=Encoder error
+void dvr_update_status() {
+    pthread_mutex_lock(&dvr_mutex);
+    if (dvr_is_recording) {
+        int ret = -1;
+        FILE *fp = fopen("/tmp/record.dat", "r");
+        if (fp) {
+            fscanf(fp, "%d", &ret);
+            fclose(fp);
+        }
+        if (ret != 1) {
+            dvr_is_recording = false;
+        }
+    }
+    pthread_mutex_unlock(&dvr_mutex);
+}
+
+void dvr_enable_line_out(bool enable) {
+    char buf[128];
+    if (enable) {
+        sprintf(buf, "%s out_on", AUDIO_SEL_SH);
+        system(buf);
+        sprintf(buf, "%s out_linein_on", AUDIO_SEL_SH);
+        system(buf);
+        sprintf(buf, "%s out_dac_off", AUDIO_SEL_SH);
+        system(buf);
+    } else {
+        sprintf(buf, "%s out_off", AUDIO_SEL_SH);
+        system(buf);
+    }
+}
+
+void dvr_select_audio_source(uint8_t source) {
+    char buf[128];
+    char *audio_source[3] = {
+        "in_mic1",
+        "in_mic2",
+        "in_linein"};
+
+    if (source > 2)
+        source = 2;
+    sprintf(buf, "%s %s", AUDIO_SEL_SH, audio_source[source]);
+    system(buf);
+}
+
+static void dvr_update_record_conf() {
+    if (g_setting.record.format_ts)
```

Comment:
We should maintain a value with the values in it so we can short circuit this function if none of the values have changed. This *may* alleviate some of the problems seen with recordings flipping back to the defaults.

Reply:
humm, while i do like the aspect of doing only as much as necessary, i have some concerns:
- if the dvr configuration reset is indeed related to writing the ini file, this will only make it more in-frequent and does not fix the root cause. we don't really want that right now, the issue is hard enough to trace as it is.
- if there are additional changes to the record parameters this could become very error prone because the developers will need to take great care that the file is indeed written out when one of the parameters change.

i vote we get this refactor merged and then investigate more. maybe take a peek at the minini write function, it does some interesting stuff with writing to a tmp file in-place and then moving it over. maybe we should swap that out for something where we have more control about when the actual file is written and, more importantly, we get notified if the write fails. right now there is zero validation whether a write actual took place or not.

Reply:
Fair enough. If the minIni library was replaced with something where we have more control, then we could have that only do the writing if the data has changed.